### PR TITLE
Previously LXD profiles were overwritten, now these are merged.

### PR DESCRIPTION
**Pull request by conloos**
##### SUMMARY
Previously LXD profiles were overwritten, now these are merged together.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lxd_profile: Profile Management

##### ADDITIONAL INFORMATION
Previously LXD profiles were overwritten, now these are merged.

So far, only configurations that were defined in Ansible have been adopted. Local changes were thus rejected. E.g. a configuration was only made in the config section, then the existing configurations in all other sections disappeared. especially in the default profile that was uncomfortable.
```

### CODE REVIEW

**Comment by felixfontein** [comment]
@conloos please keep working on one PR, instad of closing / reopening new PRs all the time. Thanks.

**Comment by felixfontein** [comment]
Two general comments:
1. Every change needs a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).
2. We value backwards compatibility, so this feature needs to be controlled by a (new?) module option, where the old behavior is the default.

**Comment by conloos** [comment]
> @conloos please keep working on one PR, instad of closing / reopening new PRs all the time. Thanks.

Hi @felixfontein,

I have revised the plugin and restored the original behavior. A new option enables the existing profile to be expanded.

The new version is in my [fork.](https://github.com/conloos/community.general) How can I incorporate these changes into this PR without closing them?

I should also create a changelog fragment. Where do I get the revision number or which is it created?
E.g .: changelogs / fragments / 1478-filesystem-fix-1457-resizefs-idempotency.yml 
The 1457, is that the ticket number in my case: 1748? 
How can I know this PR number before I have created the PR, or is the normal behavior of creating a PR and then adding the fragment? 


Is there a blog or something similar that I can use to complete the next steps without try and error?

**Comment by felixfontein** [comment]
> I should also create a changelog fragment. Where do I get the revision number or which is it created?

It's this PR's number, i.e. 1748.

> E.g .: changelogs / fragments / 1478-filesystem-fix-1457-resizefs-idempotency.yml
> The 1457, is that the ticket number in my case: 1748?

No, 1457 would be the issue number if there is one.

I'd simply call it `1748-lxd_profile-merge-profiles.yml` or something like that.

> How can I know this PR number before I have created the PR, or is the normal behavior of creating a PR and then adding the fragment?

The latter is rather normal. If you're adventurous you can always grab the largest issue/PR number directly before creating the PR from https://github.com/ansible-collections/community.general/issues/?q= and increment it by one. That's usually the correct value (if no other issue/PR got inbetween). But more common is to just add another commit with the fragment.

> Is there a blog or something similar that I can use to complete the next steps without try and error?

You can always run the sanity tests locally before pushing a commit: https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#testing-collections

Please note that you need to check out the collection repo in the correct place for this to work (see https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#contributing-to-collections for how to use modules from it as well).

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 1]
```diff
@@ -195,7 +216,7 @@
 
 # CONFIG_PARAMS is a list of config attribute names.
 CONFIG_PARAMS = [
-    'config', 'description', 'devices'
+    'config', 'description', 'devices', 'merge'
```
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_profile.py:219]: I don't think `merge` should be part of the configuration.

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 2]
```diff
@@ -307,10 +328,96 @@ def _needs_to_apply_profile_configs(self):
...
+        Args:
+            None
+        Kwargs:
+            None
+        Raises:
+            None
+        Returns:
+            None"""
+        config = self.old_profile_json.copy()
+        if 'merge' in self.config.keys() and self.config['merge'] is True:
```
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_profile.py:415]:
```suggestion
        if self.module.params['mege']:
```

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 3]
```diff
@@ -307,10 +328,96 @@ def _needs_to_apply_profile_configs(self):
...
+            dict(config): Dict with the old config in 'metadata' and new config in 'config'
+        Kwargs:
+            None
+        Raises:
+            None
+        Returns:
+            dict(config): new config"""
         for k, v in self.config.items():
             config[k] = v
+        return(config)
```
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_profile.py:395]:
```suggestion
        return config
```

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 4]
```diff
@@ -307,10 +328,96 @@ def _needs_to_apply_profile_configs(self):
...
+        Returns:
+            dict(config): new config"""
+        # merge or copy the sections from the existing profile to 'config'
+        for item in ['config', 'description', 'devices', 'name', 'used_by']:
+            if item in config:
+                config[item] = self._merge_dicts(config['metadata'][item], config[item])
+            else:
+                config[item] = config['metadata'][item]
+        # merge or copy the sections from the ansible-task to 'config'
+        return (self._merge_dicts(self.config, config))
```
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_profile.py:375]:
```suggestion
        return self._merge_dicts(self.config, config)
```

**Review by felixfontein — COMMENTED** [review]
I have some generic comments. I don't know or use the module, so I cannot comment much on whether this change does what it claims to do.

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 5]
```diff
@@ -52,6 +52,12 @@
             See U(https://github.com/lxc/lxd/blob/master/doc/rest-api.md#post-11)
         required: false
         type: str
+    merge:
+        description:
+            - Merge the configuration of the present profile and the new desired configitems
+        required: false
+        default: false
+        type: bool
```
[the remark this answers is not in the document]
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:60]: Thanks, I have included the suggestion in the code.

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 6]
```diff
@@ -52,6 +52,12 @@
             See U(https://github.com/lxc/lxd/blob/master/doc/rest-api.md#post-11)
         required: false
         type: str
+    merge:
+        description:
+            - Merge the configuration of the present profile and the new desired configitems
```
[the remark this answers is not in the document]
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:57]: Thanks, I have included the suggestion in the code.

File: plugins/modules/cloud/lxd/lxd_profile.py [hunk 7]
```diff
@@ -52,6 +52,12 @@
             See U(https://github.com/lxc/lxd/blob/master/doc/rest-api.md#post-11)
         required: false
         type: str
+    merge:
```
[the remark this answers is not in the document]
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:55]: At that point I'm emotionless, I've replaced merge with merge_profile

File: plugins/modules/cloud/lxd/lxd_profile.py (hunk 2 above)
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:415]: Thanks, I have included the suggestion in the code.

File: plugins/modules/cloud/lxd/lxd_profile.py (hunk 3 above)
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:395]: Thanks, I have included the suggestion in the code.

File: plugins/modules/cloud/lxd/lxd_profile.py (hunk 4 above)
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:375]: Thanks, I have included the suggestion in the code.

File: plugins/modules/cloud/lxd/lxd_profile.py (hunk 1 above)
Comment (conloos) [reply plugins/modules/cloud/lxd/lxd_profile.py:219]: Ok. Thanks. It works without defined in the CONFIG_PARAMS.
I have included the suggestion in the code.

File: changelogs/fragments/1748-lxd_profile-merge-profiles.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- lxd_profile - add ``merge_profile`` to merge configurations from the play to an existing profile
```
Comment (felixfontein) [inline changelogs/fragments/1748-lxd_profile-merge-profiles.yml:2]:
```suggestion
- lxd_profile - added ``merge_profile`` parameter to merge configurations from the play to an existing profile (https://github.com/ansible-collections/community.general/pull/1748).
```

**Review by felixfontein — COMMENTED** [review]
> @conloos This PR contains `@` mentions in at least one commit message. Those mentions can cause cascading notifications through GitHub and need to be removed. Please squash or amend your commits to remove the mentions.

Could you please remove the mention from the commit message?

File: changelogs/fragments/1748-lxd_profile-merge-profiles.yml (hunk 8 above)
Comment (conloos) [reply changelogs/fragments/1748-lxd_profile-merge-profiles.yml]: Thanks: its included

**Pull request closed without merging**